wandb/apis/importers/wandb.py

```diff
@@ -0,0 +1,1133 @@
+import itertools
+import json
+import time
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from dataclasses import dataclass
+from datetime import datetime as dt
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
+from unittest.mock import patch
+
+import requests
+import yaml
+
+import wandb
+from wandb.apis.public import Run
+from wandb.util import coalesce, get_module, remove_keys_with_none_values
+
+with patch("click.echo"):
+    import wandb.apis.reports as wr
+    from wandb.apis.reports import Report
+
+from . import internal, progress, protocols
+from .config import ImportConfig
+from .logs import _thread_local_settings, wandb_logger
+from .protocols import ArtifactSequence
+
+pl = get_module(
+    "polars",
+    required="Missing `polars`, try `pip install polars`",
+)
+
+rich = get_module(
+    "rich",
+    required="Missing `rich`, try `pip install rich`",
+)
+
+ART_SEQUENCE_DUMMY_DESCRIPTION = "__ART_SEQUENCE_DUMMY_DESCRIPTION__"
+
+
+class WandbRun:
+    def __init__(self, run: Run):
+        self.run = run
+        self.api = wandb.Api(
+            api_key=_thread_local_settings.api_key,
+            overrides={"base_url": _thread_local_settings.base_url},
+        )
+
+        _thread_local_settings.entity = self.entity()
+        _thread_local_settings.project = self.project()
+        _thread_local_settings.run_id = self.run_id()
+
+        # For caching
+        self._files: Optional[Iterable[Tuple[str, str]]] = None
+        self._artifacts: Optional[Iterable[wandb.Artifact]] = None
+        self._used_artifacts: Optional[Iterable[wandb.Artifact]] = None
+
+    def run_id(self) -> str:
+        return self.run.id
+
+    def entity(self) -> str:
+        return self.run.entity
+
+    def project(self) -> str:
+        return self.run.project
+
+    def config(self) -> Dict[str, Any]:
+        return self.run.config
+
+    def summary(self) -> Dict[str, float]:
+        s = self.run.summary
+
+        # Modify artifact paths because they are different between systems
+        s = self._modify_table_artifact_paths(s)
```

Comment:
There are more cases where we store a reference to an artifact, e.g.:
```
first = wandb.Artifact("first", type="my_type")
with first.new_file("first.txt") as f:
    f.write("first")
first.save()
first.wait()

second = wandb.Artifact("second", type="my_type")
second.add_reference(first.get_path("first.txt"))
second.save()
```

You can do the same in a less obvious way by creating tables using data from other tables.

Reply:
This creates a dependency between the artifacts. We could at least check whether the artifacts we're importing don't have dependencies, see [here](https://github.com/wandb/core/blob/029e1e1dcf838b432bff86a0a786a2697af578b2/services/gorilla/schema.graphql#L2330-L2341).

Reply:
I think that case is rare atm.  How would you handle it?